File: app/src/main/java/com/moisesborges/tvaddict/showdetails/ShowDetailsActivity.java
package com.moisesborges.tvaddict.showdetails;

import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Html;
import android.text.Spanned;
import android.view.View;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.moisesborges.tvaddict.App;
import com.moisesborges.tvaddict.R;
import com.moisesborges.tvaddict.adapters.CastAdapter;
import com.moisesborges.tvaddict.adapters.ItemClickListener;
import com.moisesborges.tvaddict.adapters.SeasonsAdapter;
import com.moisesborges.tvaddict.episodes.EpisodesActivity;
import com.moisesborges.tvaddict.models.CastMember;
import com.moisesborges.tvaddict.models.Externals;
import com.moisesborges.tvaddict.models.Season;
import com.moisesborges.tvaddict.models.Show;
import com.moisesborges.tvaddict.utils.ProgressBarHelper;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;

/**
 * Created by Moisés on 17/04/2017.
 */

public class ShowDetailsActivity extends AppCompatActivity implements ShowDetailsView {

    private static final String SHOW_ARG = "com.moisesborges.tvaddict.showdetails.ShowDetailsActivity.show";

    @Inject
    ShowDetailsPresenter mShowDetailsPresenter;

    private Unbinder mUnbinder;
    private Show mShow;

    @BindView(R.id.show_image_view)
    ImageView mShowImageView;
    @BindView(R.id.show_summary_text_view)
    TextView mShowSummaryTextView;
    @BindView(R.id.toolbar)
    Toolbar mToolbar;
    @BindView(R.id.seasons_recycler_view)
    RecyclerView mSeasonsRecyclerView;
    @BindView(R.id.seasons_progress_bar)
    ProgressBar mSeasonsProgressBar;
    @BindView(R.id.cast_recycler_view)
    RecyclerView mCastRecyclerView;
    @BindView(R.id.cast_progress_bar)
    ProgressBar mCastProgressBar;
    @BindView(R.id.save_show_float_action_button)
    FloatingActionButton mFloatingActionButton;
    @BindView(R.id.show_genres_text_view)
    TextView mGenresTextView;
    @BindView(R.id.show_network_text_view)
    TextView mNetworkTextView;
    @BindView(R.id.show_runtime_text_view)
    TextView mRuntimeTextView;
    @BindView(R.id.show_ration_text_view)
    TextView mRatingTextView;

    private ItemClickListener<Season> mSeasonItemClickListener = (season) -> mShowDetailsPresenter.openEpisodes(mShow, season);

    private SeasonsAdapter mSeasonsAdapter = new SeasonsAdapter(mSeasonItemClickListener);

    private CastAdapter mCastAdapter = new CastAdapter();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_details);
        mUnbinder = ButterKnife.bind(this);
        setupToolbar();
        setupRecyclerViews();
        injectDependencies();
        mShow = getIntent().getParcelableExtra(SHOW_ARG);
    }

    private void setupRecyclerViews() {
        mSeasonsRecyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        mSeasonsRecyclerView.setHasFixedSize(true);
        mSeasonsRecyclerView.setAdapter(mSeasonsAdapter);

        mCastRecyclerView.setLayoutManager(new GridLayoutManager(this, 3));
        mCastRecyclerView.setHasFixedSize(true);
        mCastRecyclerView.setAdapter(mCastAdapter);
    }

    private void setupToolbar() {
        setSupportActionBar(mToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    @Override
    protected void onStart() {
        super.onStart();
        mShowDetailsPresenter.bindView(this);
        mShowDetailsPresenter.loadShowDetails(mShow);
    }

    @Override
    protected void onStop() {
        super.onStop();
        mShowDetailsPresenter.unbindView();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mUnbinder.unbind();
    }

    private void injectDependencies() {
        ((App) getApplicationContext()).getAppComponent().inject(this);
    }

    public static void start(@NonNull Context context, @NonNull Show show) {
        Intent intent = new Intent(context, ShowDetailsActivity.class);
        intent.putExtra(SHOW_ARG, show);
        context.startActivity(intent);
    }

    @OnClick(R.id.save_show_float_action_button)
    public void onSaveShowClick(View view) {
        mShowDetailsPresenter.changeWatchingStatus(mShow);
    }

    @Override
    public void setShowName(String showName) {
        this.setTitle(showName);
    }

    @Override
    public void setShowImage(String imageUrl) {
        Glide.with(this)
                .load(imageUrl)
                .into(mShowImageView);
    }

    @Override
    public void setShowSummary(String summary) {
        mShowSummaryTextView.setText(extractFromHtml(summary));
    }

    @Override
    public void displaySeasons(@NonNull List<Season> seasons) {
        mSeasonsAdapter.setSeasons(seasons);
    }

    @Override
    public void displayAdditionalInfoLoadingInProgress(boolean isLoading) {
        ProgressBarHelper.show(mSeasonsProgressBar, isLoading);
        ProgressBarHelper.show(mCastProgressBar, isLoading);
    }

    @Override
    public void displayAdditionalInfoNotLoaded(boolean hasError) {

    }

    @Override
    public void setShow(@NonNull Show show) {
        mShow = show;
    }

    @Override
    public void navigateToEpisodes(Show show, int seasonNumber) {
        EpisodesActivity.start(this, show, seasonNumber);
    }

    @Override
    public void displaySavedShowMessage() {
        showToast(R.string.show_saved_with_success);
    }

    private void showToast(@StringRes int message) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void displayCastMembers(List<CastMember> castMembers) {
        mCastAdapter.setCast(castMembers);
    }

    @Override
    public void displaySaveShowButton(boolean shouldDisplaySavedAction) {
        mFloatingActionButton.setImageResource(shouldDisplaySavedAction ?
                R.drawable.ic_add_white_24px : R.drawable.ic_clear_white_24px);
        mFloatingActionButton.setEnabled(true);
    }

    @Override
    public void displayShowRemovedMessage() {
        showToast(R.string.show_removed_with_success);
    }

    @Override
    public void setShowRating(Double rating) {
        String ratingLabel = rating != null ?
                rating.toString() : getResources().getString(R.string.not_available);
        mRatingTextView.setText(ratingLabel);
    }

    @Override
    public void setShowRuntime(Integer runtime) {
        String runtimeLabel = runtime != null ?
                getResources().getString(R.string.runtime_label, runtime) : getResources().getString(R.string.not_available);
        mRuntimeTextView.setText(runtimeLabel);
    }

    @Override
    public void setShowNetwork(String networkName) {
        mNetworkTextView.setText(networkName);
    }

    @Override
    public void setShowGenres(List<String> genres) {
        mGenresTextView.setText(listOfGenres(genres));
    }

    private String listOfGenres(List<String> genres) {
        StringBuilder builder = new StringBuilder();
        for (int pos = 0; pos < genres.size(); pos++) {
            if (pos != 0) {
                builder.append(" - ");
            }
            builder.append(genres.get(pos));
        }
        return builder.toString();
    }

    @Override
    public void setShowExternalLinks(Externals externals) {

    }

    private Spanned extractFromHtml(String html) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            return Html.fromHtml(html, Html.FROM_HTML_MODE_COMPACT);
        } else {
            return Html.fromHtml(html);
        }
    }


}

File: app/src/main/java/com/moisesborges/tvaddict/showdetails/ShowDetailsActivity.kt
package com.moisesborges.tvaddict.showdetails

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.annotation.StringRes
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.widget.Toast
import com.bumptech.glide.Glide
import com.moisesborges.tvaddict.App
import com.moisesborges.tvaddict.R
import com.moisesborges.tvaddict.adapters.CastAdapter
import com.moisesborges.tvaddict.adapters.ItemClickListener
import com.moisesborges.tvaddict.adapters.SeasonsAdapter
import com.moisesborges.tvaddict.episodes.EpisodesActivity
import com.moisesborges.tvaddict.models.CastMember
import com.moisesborges.tvaddict.models.Externals
import com.moisesborges.tvaddict.models.Season
import com.moisesborges.tvaddict.models.Show
import com.moisesborges.tvaddict.utils.HtmlUtils
import com.moisesborges.tvaddict.utils.ProgressBarHelper
import kotlinx.android.synthetic.main.activity_show_details.*
import kotlinx.android.synthetic.main.content_main_show_details.*
import javax.inject.Inject

/**
 * Created by Moisés on 17/04/2017.
 */

class ShowDetailsActivity : AppCompatActivity(), ShowDetailsView {

    @Inject
    internal lateinit var showDetailsPresenter: ShowDetailsPresenter

    private lateinit var show: Show


    private val seasonItemClickListener = ItemClickListener<Season> { season -> showDetailsPresenter.openEpisodes(show, season) }

    private val seasonsAdapter = SeasonsAdapter(seasonItemClickListener)

    private val castAdapter = CastAdapter()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_show_details)
        setupToolbar()
        setupRecyclerViews()
        injectDependencies()
        show = intent.getParcelableExtra(SHOW_ARG)
        save_show_float_action_button.setOnClickListener { showDetailsPresenter.changeWatchingStatus(show) }
    }

    private fun setupRecyclerViews() {
        seasons_recycler_view.layoutManager = GridLayoutManager(this, 3)
        seasons_recycler_view.setHasFixedSize(true)
        seasons_recycler_view.adapter = seasonsAdapter

        cast_recycler_view.layoutManager = LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false)
                //GridLayoutManager(this, 3)
        cast_recycler_view.setHasFixedSize(true)
        cast_recycler_view.adapter = castAdapter
    }

    private fun setupToolbar() {
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        title = ""
    }

    override fun onStart() {
        super.onStart()
        showDetailsPresenter.bindView(this)
        showDetailsPresenter.loadShowDetails(show)
    }

    override fun onStop() {
        super.onStop()
        showDetailsPresenter.unbindView()
    }

    override fun onDestroy() {
        super.onDestroy()
    }

    private fun injectDependencies() {
        (applicationContext as App).appComponent.inject(this)
    }

    override fun setShowName(showName: String) {
        show_name_text_view.text = showName
        toolbar.title = showName
    }

    override fun setShowImage(imageUrl: String) {
        Glide.with(this)
                .load(imageUrl)
                .into(show_image_view)
    }

    override fun setShowSummary(summary: String) {
        show_summary_text_view.text = HtmlUtils.extractFromHtml(summary)
    }

    override fun displaySeasons(seasons: List<Season>) {
        seasonsAdapter.setSeasons(seasons)
    }

    override fun displayAdditionalInfoLoadingInProgress(isLoading: Boolean) {
        ProgressBarHelper.show(seasons_progress_bar, isLoading)
        ProgressBarHelper.show(cast_progress_bar, isLoading)
    }

    override fun displayAdditionalInfoNotLoaded(hasError: Boolean) {

    }

    override fun setShow(show: Show) {
        this.show = show
    }

    override fun navigateToEpisodes(show: Show, seasonNumber: Int) {
        EpisodesActivity.start(this, show, seasonNumber)
    }

    override fun displaySavedShowMessage() {
        showToast(R.string.show_saved_with_success)
    }

    private fun showToast(@StringRes message: Int) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    override fun displayCastMembers(castMembers: List<CastMember>) {
        castAdapter.setCast(castMembers)
    }

    override fun displaySaveShowButton(shouldDisplaySavedAction: Boolean) {
        save_show_float_action_button.setImageResource(if (shouldDisplaySavedAction)
            R.drawable.ic_add_white_24px
        else
            R.drawable.ic_clear_white_24px)
        save_show_float_action_button.isEnabled = true
    }

    override fun displayShowRemovedMessage() {
        showToast(R.string.show_removed_with_success)
    }

    override fun setShowRating(rating: Double?) {
        val ratingLabel = rating?.toString() ?: resources.getString(R.string.not_available)
        show_rating_text_view.text = ratingLabel
    }

    override fun setShowRuntime(runtime: Int?) {
        val runtimeLabel = if (runtime != null)
            resources.getString(R.string.runtime_label, runtime)
        else
            resources.getString(R.string.not_available)
        show_runtime_text_view.text = runtimeLabel
    }

    override fun setShowNetwork(networkName: String) {
        show_network_text_view.text = networkName
    }

    override fun setShowGenres(genres: List<String>) {
        show_genres_text_view.text = listOfGenres(genres)
    }

    private fun listOfGenres(genres: List<String>): String {
        val builder = StringBuilder()
        for (pos in genres.indices) {
            if (pos != 0) {
                builder.append(", ")
            }
            builder.append(genres[pos])
        }
        return builder.toString()
    }

    override fun setShowExternalLinks(externals: Externals) {

    }

    companion object {

        private val SHOW_ARG = "com.moisesborges.tvaddict.showdetails.ShowDetailsActivity.show"

        @JvmStatic
        fun start(context: Context, show: Show) {
            val intent = Intent(context, ShowDetailsActivity::class.java)
            intent.putExtra(SHOW_ARG, show)
            context.startActivity(intent)
        }
    }


}
